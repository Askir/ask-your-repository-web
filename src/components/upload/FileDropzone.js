// @flow
import React from 'react';
import Dropzone from 'react-dropzone';
import type { DropFilesEventHandler } from 'react-dropzone';
import './FileDropzone.scss';

type Props = {
  onDrop: DropFilesEventHandler,
};

function FileDropzone(props: Props) {
  const { onDrop } = props;

  return (
    <Dropzone
      multiple={false}
      className="FileDropzone"
      activeClassName="FileDropzone--active"
      acceptClassName="FileDropzone--accept"
      disabledClassName="FileDropzone--disabled"
      onDrop={onDrop}
      data-cy="upload-dropzone"
    >
      <span className="FileDropzone__text">
        Place your file here
      </span>
    </Dropzone>
  );
}

export default FileDropzone;
