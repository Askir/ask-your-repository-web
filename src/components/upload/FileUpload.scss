@import '../../style/breakpoints.scss';

.FileUpload {
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 90vh;
  width: 100%;
  padding: 30px;

  @media (min-width: $xs-to-sm) {
    height: auto;
  }
}
