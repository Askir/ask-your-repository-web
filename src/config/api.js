import axios from 'axios';
import qs from 'qs';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  requestInterceptor: config => ({ ...config }),
  paramsSerializer: parameters => qs.stringify(parameters, { arrayFormat: 'repeat' }),
});

export default api;
