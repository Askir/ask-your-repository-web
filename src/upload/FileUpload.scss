.FileUpload {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 30px 0;

  &__image-preview {
    margin: 20px 0;
    width: 800px;
    border-radius: 5px;
    box-shadow: 0 4px 10px 5px rgba(0, 0, 0, .06);
  }
}
