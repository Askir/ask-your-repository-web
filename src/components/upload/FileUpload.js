// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { flashErrorMessage } from 'redux-flash';
import { Redirect } from 'react-router-dom';
import FileDropzone from './FileDropzone';
import ActivityIndicator from '../utility/ActivityIndicator';
import ImageService from '../../services/ImageService';
import { setImage } from '../../state/image/image.actionCreators';
import type { Team } from '../../models/Team';
import type { AppState } from '../../state/AppState';
import './FileUpload.scss';
import { active } from 'glamor';

type Props = {
  dispatch: Function,
  activeTeam: ?Team,
};

type State = {
  isUploading: boolean,
  hasUploaded: boolean,
};

class FileUpload extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isUploading: false,
      hasUploaded: false,
    };
  }

  handleImageDrop = (files: Blob[]) => {
    if (files && files[0]) {
      const image = files[0];
      this.submitImage(image);
    }
  };

  async submitImage(imageData: Blob) {
    const { dispatch, activeTeam } = this.props;

    if (!activeTeam) return;

    const formData = new FormData();
    formData.append('image', imageData);
    formData.append('team_id', activeTeam.id);

    this.setState({ isUploading: true });

    try {
      const image = await ImageService.create(formData);
      dispatch(setImage(image));
      this.setState({ hasUploaded: true });
    } catch (error) {
      const message = error.response
        ? error.response.message
        : 'Our server seems to be unavailable at the moment.';
      dispatch(flashErrorMessage(message));
    }

    this.setState({ isUploading: false });
  }

  render() {
    if (this.state.hasUploaded) {
      return <Redirect to="/tagging" />;
    }

    const { isUploading } = this.state;

    return (
      <div className="FileUpload">
        <div className="FileUpload__inner">
          {isUploading ? (
            <ActivityIndicator text="Uploading" />
          ) : (
            <FileDropzone onDrop={this.handleImageDrop} />
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
});

export default connect(mapStateToProps)(FileUpload);
