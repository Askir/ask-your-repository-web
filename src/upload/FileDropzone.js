// @flow
import React from 'react';
import Dropzone from 'react-dropzone';
import type { DropFilesEventHandler } from 'react-dropzone';
import './FileDropzone.scss';

type Props = {
  onDrop: DropFilesEventHandler,
};

function FileDropzone(props: Props) {
  const { onDrop } = props;

  return (
    <Dropzone
      multiple={false}
      className="Dropzone"
      activeClassName="Dropzone--active"
      acceptClassName="Dropzone--accept"
      disabledClassName="Dropzone--disabled"
      onDrop={onDrop}
    >
      Place your file here
    </Dropzone>
  );
}

export default FileDropzone;
