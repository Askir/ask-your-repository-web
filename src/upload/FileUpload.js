import React, { Component } from 'react';
import UploadDropzone from './UploadDropzone';
import './FileUpload.scss';
import Message from '../utility/Message';
import ErrorMessage from '../utility/ErrorMessage';
import SuccessMessage from '../utility/SuccessMessage';
import fetchUploadImage from '../requests/imageRequests';

class FileUpload extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedImage: null,
      isUploading: false,
      hasUploaded: false,
      uploadError: null,
    };
  }

  handleFileReaderLoad = (e) => {
    const { result } = e.target;
    this.setState({ selectedImage: result });
  };

  handleImageDrop = async (files) => {
    if (files && files[0]) {
      const image = files[0];
      await this.submitImage(image);

      if (this.state.hasUploaded) {
        const reader = new FileReader();
        reader.onload = this.handleFileReaderLoad;
        reader.readAsDataURL(image);
      }
    }
  };

  async submitImage(image) {
    const formData = new FormData();
    formData.append('image', image);

    this.setState({ isUploading: true });

    fetchUploadImage(formData)
      .then(() => this.setState({ hasUploaded: true }))
      .catch((error) => {
        if (error.response) {
          this.setState({ uploadError: error.response.message });
        } else {
          this.setState({ uploadError: 'Our server seems to be unavailable at the moment.' });
        }
      });

    this.setState({ isUploading: false });
  }

  render() {
    const {
      selectedImage,
      isUploading,
      hasUploaded,
      uploadError,
    } = this.state;

    return (
      <div className="FileUpload">
        {isUploading && (
          <Message>
            Your image is being uploaded...
          </Message>
        )}
        {hasUploaded && (
          <SuccessMessage>
            Your image has uploaded successfully!
          </SuccessMessage>
        )}
        {uploadError && (
          <ErrorMessage>{uploadError}</ErrorMessage>
        )}
        {selectedImage ? (
          <img
            src={selectedImage}
            alt=""
            className="FileUpload__image-preview"
          />
        ) : (
          <UploadDropzone onDrop={this.handleImageDrop} />
        )}
      </div>
    );
  }
}

export default FileUpload;
