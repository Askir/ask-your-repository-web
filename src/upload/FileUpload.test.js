// @flow
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import FileUpload from './FileUpload';
import FileDropzone from './FileDropzone';
import api from '../config/api';

jest.mock('../config/api', () => ({
  post: jest.fn(() => Promise.resolve()),
}));

describe('<FileUpload />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<FileUpload />);
  });

  afterEach(() => {
    api.post.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders an FileDropzone component', () => {
    expect(wrapper.find(FileDropzone)).toBeTruthy();
  });

  describe('an image has been dropped', () => {
    const imageBlob = new Blob([], { type: 'image/jpeg' });

    beforeEach(async () => {
      await wrapper.instance().handleImageDrop([imageBlob]);
    });

    it('should send a POST request to the server', () => {
      expect(api.post).toHaveBeenCalled();
    });

    it('should set the correct Content-Type for the request', () => {
      const { headers } = api.post.mock.calls[0][2];
      expect(headers['Content-Type']).toEqual('multipart/form-data');
    });

    it('should send the dropped file as form-data', async () => {
      const formData = api.post.mock.calls[0][1];
      const sentFile = formData.get('image');

      let sentFileData: string;
      let imageBlobData: string;

      function asyncFileReader(file: Blob, onLoad: (string) => void) {
        return new Promise((resolve) => {
          const fileReader = new FileReader();
          fileReader.onload = (e) => {
            onLoad(e.target.result);
            resolve();
          };
          fileReader.readAsDataURL(file);
        });
      }

      await Promise.all([
        asyncFileReader(sentFile, (result) => { sentFileData = result; }),
        asyncFileReader(imageBlob, (result) => { imageBlobData = result; }),
      ]);

      expect(sentFileData).toEqual(imageBlobData);
    });

    describe('upload succeeded', () => {
      it('should display a success message', () => {
        expect(wrapper.text()).toEqual(expect.stringContaining('success'));
      });
    });

    describe('upload failed because of network error', () => {
      beforeAll(() => {
        api.post.mockImplementation(() => Promise.reject(Response.error()));
      });

      it('should display a network error message', () => {
        expect(wrapper.text()).toEqual(expect.stringContaining('unavailable'));
      });
    });

    describe('upload failed because of server error', () => {
      const errorMessage = 'Dummy Error';
      const errorResponse = { message: errorMessage };

      beforeAll(() => {
        api.post.mockImplementation(() => Promise.reject({ response: errorResponse }));
      });

      it('should display the error message', () => {
        expect(wrapper.text()).toEqual(expect.stringContaining(errorMessage));
      });
    });
  });

  it('do nothing if image drop is called without any files', () => {
    wrapper.instance().handleImageDrop(null);
    expect(api.post).not.toHaveBeenCalled();
  });
});
