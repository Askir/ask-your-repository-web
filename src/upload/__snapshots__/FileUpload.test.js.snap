// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUpload /> renders correctly 1`] = `
<FileUpload>
  <div
    className="FileUpload"
  >
    <UploadDropzone
      onDrop={[Function]}
    >
      <n
        accept="image/*"
        acceptClassName="Dropzone--accept"
        activeClassName="Dropzone--active"
        className="Dropzone"
        disableClick={false}
        disabled={false}
        disabledClassName="Dropzone--disabled"
        getDataTransferItems={[Function]}
        inputProps={Object {}}
        maxSize={Infinity}
        minSize={0}
        multiple={false}
        onDrop={[Function]}
        preventDropOnDocument={true}
      >
        <div
          aria-disabled={false}
          className="Dropzone"
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          Place your file here
          <input
            accept="image/*"
            autoComplete="off"
            disabled={false}
            multiple={false}
            onChange={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 0.00001,
                "pointerEvents": "none",
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            type="file"
          />
        </div>
      </n>
    </UploadDropzone>
  </div>
</FileUpload>
`;
